zone_affinity.go

```diff
@@ -0,0 +1,169 @@
+// where the leader is in the same zone.  Doing so can help from a performance
+// perspective by minimizing round trip latency, and it can also help keep costs
+// down when running in AWS by avoid cross-AZ data transfer.  Partitions are
+// spread evenly across consumers.
+//
+// The consumers communicate their zones via UserData.  If the zone is missing
+// or cannot be determined, then the data will be empty.  The zone can also be
+// set manually using the AVAILABILITY_ZONE environment variable...this is
+// useful e.g. when doing local testing.
+type ZoneAffinityBalancer struct {
+	zone []byte
+}
+
+func NewZoneAffinityBalancer() *ZoneAffinityBalancer {
+	return &ZoneAffinityBalancer{zone: findZone()}
+}
+
+func (za *ZoneAffinityBalancer) Balance(members []Member, partitions []Partition) map[string][]int32 {
+	zonedPartitions := make(map[string][]int32)
+	for _, part := range partitions {
+		zone := part.Leader.Rack()
+		zonedPartitions[zone] = append(zonedPartitions[zone], part.ID)
+	}
+
+	zonedConsumers := make(map[string][]string)
+	for _, member := range members {
+		zone := string(member.UserData)
+		zonedConsumers[zone] = append(zonedConsumers[zone], member.ID)
+	}
+
+	targetPerMember := len(partitions) / len(members)
+	remainder := len(partitions) % len(members)
+	assignments := make(map[string][]int32)
+
+	// assign as many as possible in zone.  this will assign up to partsPerMember
+	// to each consumer.  it will also prefer to allocate remainder partitions
+	// in zone if possible.
+	for zone, parts := range zonedPartitions {
+
+		consumers := zonedConsumers[zone]
+		if len(consumers) == 0 {
+			continue
+		}
+
+		// don't over-allocate.  cap partition assignments at the calculated
+		// target.
+		partsPerMember := len(parts) / len(consumers)
+		if partsPerMember > targetPerMember {
+			partsPerMember = targetPerMember
+		}
+
+		for _, consumer := range consumers {
+			assignments[consumer] = append(assignments[consumer], parts[:partsPerMember]...)
+			parts = parts[partsPerMember:]
+		}
+
+		// if we had enough partitions for each consumer in this zone to hit its
+		// target, attempt to use any leftover partitions to satisfy the total
+		// remainder by adding at most 1 partition per consumer.
+		leftover := len(parts)
+		if partsPerMember == targetPerMember {
+			if leftover > remainder {
+				leftover = remainder
+			}
+			if leftover > len(consumers) {
+				leftover = len(consumers)
+			}
+			remainder -= leftover
+		}
+
+		// this loop covers the case where we're assigning extra partitions or
+		// if there weren't enough to satisfy the targetPerMember and the zoned
+		// partitions didn't divide evenly.
+		for i := 0; i < leftover; i++ {
+			assignments[consumers[i]] = append(assignments[consumers[i]], parts[i])
+		}
+		parts = parts[leftover:]
+
+		if len(parts) == 0 {
+			delete(zonedPartitions, zone)
+		} else {
+			zonedPartitions[zone] = parts
+		}
+	}
+
+	// assign out remainders regardless of zone.
```

Comment:
Might be cool for a metric here for assignments any zone `stats.Incr('zab.balance.anyzone', count)`

Reply:
Yeah...I was thinking about stats.  I kind of doubt we're going to open a PR with sarama-cluster (esp. since they're deprecating it now that there's CG in sarama), so it's probably okay to start segment-ifying this code.